jscomp/ml/bigint_utils.ml

```diff
@@ -0,0 +1,71 @@
+(* 
+  Removes leading zeros from the string only if the first non-zero character
+  encountered is a digit. Unlike int and float, bigint cannot be of_string, so
+  This function removes only leading 0s. Instead, values like 00x1 are not converted
+  and are intended to be syntax errors.
+
+  000n -> 0n
+  001n -> 1n
+  01_000_000n -> 1000000n
+  0x1 -> 0x1n
+  0x001n -> 0x001n
+  -00100n -> -100n
+  
+  The following values are syntax errors
+
+  00o1n -> 00o1n 
+  00x1_000_000n -> 00x1000000n
+*)
+let remove_leading_zeros str =
+  let aux str =
+    let len = String.length str in
+    if len = 0 then ""
+    else
+      let is_digit c = c >= '0' && c <= '9' in
+      let idx = ref 0 in
+      while !idx < len && str.[!idx] = '0' do
+        incr idx
+      done;
+      if !idx >= len then "0" (* If the string contains only '0's, return '0'. *)
+      else if (is_digit str.[!idx]) then String.sub str !idx (len - !idx) (* Remove leading zeros and return the rest of the string. *)
+      else str
+  in
+  (* Replace the delimiters '_' inside number *)
+  let str = String.concat "" (String.split_on_char '_' str) in
+  (* Check if negative *)
+  let starts_with_minus = str <> "" && str.[0] = '-' in
+  let str = if starts_with_minus then String.sub str 1 (String.length str - 1) else str in
+  let processed_str = aux str in
+  if starts_with_minus then "-" ^ processed_str else processed_str
+
+let is_numeric s =
+  let len = String.length s in
+  if len = 0 then false
+  else
+    let is_digit c = (c >= '0' && c <= '9') || c = '_' in
+    let first_char = s.[0] in
+    if first_char <> '-' && not (is_digit first_char) then false
+    else
+      let rec check idx =
+        if idx >= len then true
+        else
+          let c = s.[idx] in
+          if is_digit c then check (idx + 1)
+          else false
+      in
+      check 1
+
+let compare s0 s1 =
+  (* check if negative *)
+  let is_neg s = String.length s > 0 && s.[0] = '-' in
+  match (is_neg s0, is_neg s1) with
+  | (true, false) -> -1  (* If only s0 is negative, it's smaller. *)
+  | (false, true) -> 1   (* If only s1 is negative, s0 is larger. *)
+  | _ ->
+    (* If both numbers are either negative or positive, compare their lengths. *)
+    let len0, len1 = (String.length s0, String.length s1) in
+    if len0 = len1 then String.compare s0 s1  (* If lengths are equal, compare the strings directly. *)
```

Comment:
this seems wrong for negative numbers

Reply:
It seems fine to me. Can you elaborate the case wrong?
https://ocaml.org/play#code=bGV0IGNvbXBhcmUgczAgczEgPQogICgqIGNoZWNrIGlmIG5lZ2F0aXZlICopCiAgbGV0IGlzX25lZyBzID0gU3RyaW5nLmxlbmd0aCBzID4gMCAmJiBzLlswXSA9ICctJyBpbgogIG1hdGNoIChpc19uZWcgczAsIGlzX25lZyBzMSkgd2l0aAogIHwgKHRydWUsIGZhbHNlKSAtPiAtMSAgKCogSWYgb25seSBzMCBpcyBuZWdhdGl2ZSwgaXQncyBzbWFsbGVyLiAqKQogIHwgKGZhbHNlLCB0cnVlKSAtPiAxICAgKCogSWYgb25seSBzMSBpcyBuZWdhdGl2ZSwgczAgaXMgbGFyZ2VyLiAqKQogIHwgXyAtPgogICAgKCogSWYgYm90aCBudW1iZXJzIGFyZSBlaXRoZXIgbmVnYXRpdmUgb3IgcG9zaXRpdmUsIGNvbXBhcmUgdGhlaXIgbGVuZ3Rocy4gKikKICAgIGxldCBsZW4wLCBsZW4xID0gKFN0cmluZy5sZW5ndGggczAsIFN0cmluZy5sZW5ndGggczEpIGluCiAgICBpZiBsZW4wID0gbGVuMSB0aGVuIFN0cmluZy5jb21wYXJlIHMwIHMxICAoKiBJZiBsZW5ndGhzIGFyZSBlcXVhbCwgY29tcGFyZSB0aGUgc3RyaW5ncyBkaXJlY3RseS4gKikKICAgIGVsc2UgaWYgbGVuMCA%2BIGxlbjEgdGhlbiAKICAgICAgaWYgaXNfbmVnIHMwIHRoZW4gLTEgZWxzZSAxICAoKiBBIGxvbmdlciBzMCBtZWFucyBpdCdzIGxhcmdlciB1bmxlc3MgaXQncyBuZWdhdGl2ZS4gKikKICAgIGVsc2UgICgqIGxlbjAgPCBsZW4xICopCiAgICAgIGlmIGlzX25lZyBzMSB0aGVuIDEgZWxzZSAtMSAgKCogQSBsb25nZXIgczEgbWVhbnMgczAgaXMgc21hbGxlciB1bmxlc3MgczEgaXMgbmVnYXRpdmUuICopCgpsZXQgKCkgPQooKiBsZW4wIDwgbGVuMSAgKikKICBhc3NlcnQoKGNvbXBhcmUgIi0xIiAiLTEwIiA9IDEpKTsKICBhc3NlcnQoKGNvbXBhcmUgIjEiICItMTAiID0gMSkpOwogIGFzc2VydCgoY29tcGFyZSAiLTEiICIxMCIgPSAtMSkpOwooKiBsZW4wID4gbGVuMSAgKikKICBhc3NlcnQoKGNvbXBhcmUgIi0xMDAiICItMTAiID0gLTEpKTsKICBhc3NlcnQoKGNvbXBhcmUgIjEwMCIgIi0xMCIgPSAxKSk7CiAgYXNzZXJ0KChjb21wYXJlICItMTAwIiAiMTAiID0gLTEpKTsKKCogbGVuMCA9IGxlbjEgICopCiAgYXNzZXJ0KChjb21wYXJlICItMTAiICItMTAiID0gMCkpOwogIGFzc2VydCgoY29tcGFyZSAiMTAiICItMTAiID0gMSkpOwogIGFzc2VydCgoY29tcGFyZSAiLTEwIiAiMTAiID0gLTEpKTsKICBwcmludF9lbmRsaW5lICJ0ZXN0IHBhc3NpbmchIg%3D%3D

Reply:
Ah, I figure it out this case is wrong `assert((compare "-10" "-12" = 1))` I'll fix it.

Reply:
https://github.com/rescript-lang/rescript-compiler/pull/6670/commits/84eecd7d365e5b0c119c9883ede8eb9611140b7e